domain/port/service/service.go

```diff
@@ -312,3 +237,182 @@ func verifyNoPortRangeConflicts(rangesA, rangesB []network.PortRange) error {
 	}
 	return errors.Errorf("%w: %s", port.ErrPortRangeConflict, conflicts)
 }
+
+// verifyNoColocatedPortRangeConflicts verifies the provided port ranges do not
+// conflict with the port ranges opened on units co-located with the given unit.
+func (s *Service) verifyNoColocatedPortRangeConflicts(
+	ctx domain.AtomicContext, unitUUID string, portRanges []network.PortRange,
+) error {
+	colocatedOpened, err := s.st.GetColocatedOpenedPorts(ctx, unitUUID)
+	if err != nil {
+		return errors.Errorf("failed to get opened ports co-located with unit %s: %w", unitUUID, err)
+	}
+	return verifyNoPortRangeConflicts(portRanges, colocatedOpened)
+}
```

Comment:
I don't like the fact that a service method knows about an atomic context. I'm ok with functions knowing about these things, but not methods.

Also, I'm unsure why the error can't be returned at the state layer.